### PR TITLE
Framework for Benchmarking Partial Evaluation

### PR DESCRIPTION
**Pull request by flortsch**
This PR is part of work which adds a framework for benchmarking partial evaluation in the Truffle compiler. The framework provides a parameterizable base class which benchmarks methods of the class PartialEvaluator. Sub classes implement a method to define call targets which are used during partial evaluation. It is intended that future benchmarks will use a specific builder class to create various kinds of call targets in a specific Truffle language implementation in order to benchmark corner cases of partial evaluation. The language implementation is also provided in this PR and has the following simple semantics:
- one single kind of arithmetic (addition)
- long and string literals
- if and while constructs
- read and write of local variables

### CODE REVIEW

**Comment by flortsch** [comment]
PELang, the language built in this PR for benchmarking partial evaluation, was extended to also support basic blocks and block dispatches.

**Comment by chumer** [comment]
Did you deliberately decide to not include support calls for these benchmarks? Could be interesting to see as inlining is done during PE. (depending on the strategy you use for inlining)

File: compiler/src/org.graalvm.compiler.truffle.compiler.benchmark/src/org/graalvm/compiler/truffle/compiler/benchmark/pelang/PELangIfNode.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        this.conditionNode = conditionNode;
+        this.thenNode = thenNode;
+        this.elseNode = elseNode;
+    }
+
+    @Override
+    public Object executeGeneric(VirtualFrame frame) {
+        long conditionResult = conditionNode.evaluateCondition(frame);
+
+        if (conditionResult == 0L) {
```
Comment (chumer) [inline compiler/src/org.graalvm.compiler.truffle.compiler.benchmark/src/org/graalvm/compiler/truffle/compiler/benchmark/pelang/PELangIfNode.java]: Normal languages would have a counting condition profile here. Not sure if you want to be close to languages for this.

File: compiler/src/org.graalvm.compiler.truffle.compiler/src/org/graalvm/compiler/truffle/compiler/PartialEvaluator.java [hunk 2]
```diff
@@ -231,6 +231,47 @@ public StructuredGraph createGraph(DebugContext debug, final CompilableTruffleAS
         return graph;
     }
 
+    @SuppressWarnings({"try", "unused"})
```
Comment (chumer) [inline compiler/src/org.graalvm.compiler.truffle.compiler/src/org/graalvm/compiler/truffle/compiler/PartialEvaluator.java]: Doesn't look it was necessary to move this method. Please put it back if possible.

File: compiler/src/org.graalvm.compiler.truffle.compiler.benchmark/src/org/graalvm/compiler/truffle/compiler/benchmark/pelang/PELangExpressionBlockNode.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+public final class PELangExpressionBlockNode extends PELangExpressionNode {
+
+    @Children private final PELangExpressionNode[] bodyNodes;
+
+    public PELangExpressionBlockNode(PELangExpressionNode[] bodyNodes) {
+        this.bodyNodes = bodyNodes;
+    }
+
+    @Override
+    public Object executeGeneric(VirtualFrame frame) {
```
Comment (chumer) [inline compiler/src/org.graalvm.compiler.truffle.compiler.benchmark/src/org/graalvm/compiler/truffle/compiler/benchmark/pelang/PELangExpressionBlockNode.java]: `@ExplodeLoop` is missing here.

File: compiler/src/org.graalvm.compiler.truffle.compiler.benchmark/src/org/graalvm/compiler/truffle/compiler/benchmark/pelang/PELangExpressionNode.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.nodes.UnexpectedResultException;
+
+@TypeSystemReference(PELangTypes.class)
+public abstract class PELangExpressionNode extends Node {
+
+    public abstract Object executeGeneric(VirtualFrame frame);
+
+    public long executeLong(VirtualFrame frame) throws UnexpectedResultException {
+        return PELangTypesGen.expectLong(executeGeneric(frame));
```
Comment (chumer) [inline compiler/src/org.graalvm.compiler.truffle.compiler.benchmark/src/org/graalvm/compiler/truffle/compiler/benchmark/pelang/PELangExpressionNode.java]: Doesn't look its worth using a typesystem to generate these to expectLong methods. I'd just remove the typesystem and inline its implementation here.

**Comment by flortsch** [comment]
Thanks @chumer for your feedback. It's highly appreciated. Could you elaborate on what you mean by support calls? The Graal/Truffle platform is still very new to me as I am doing this in a project for my master's. And you are right, the framework and the benchmarks will be further developed.

File: compiler/src/org.graalvm.compiler.truffle.compiler/src/org/graalvm/compiler/truffle/compiler/PartialEvaluator.java (hunk 2 above)
Comment (flortsch) [reply compiler/src/org.graalvm.compiler.truffle.compiler/src/org/graalvm/compiler/truffle/compiler/PartialEvaluator.java]: I thought about moving it to the other public methods, as I had to change the modified in order to access the method for benchmarking purposes. But I will move it back if required.

File: compiler/src/org.graalvm.compiler.truffle.compiler.benchmark/src/org/graalvm/compiler/truffle/compiler/benchmark/pelang/PELangExpressionBlockNode.java (hunk 3 above)
Comment (flortsch) [reply compiler/src/org.graalvm.compiler.truffle.compiler.benchmark/src/org/graalvm/compiler/truffle/compiler/benchmark/pelang/PELangExpressionBlockNode.java]: Thanks for spotting this.

File: compiler/src/org.graalvm.compiler.truffle.compiler.benchmark/src/org/graalvm/compiler/truffle/compiler/benchmark/pelang/PELangExpressionNode.java (hunk 4 above)
Comment (flortsch) [reply compiler/src/org.graalvm.compiler.truffle.compiler.benchmark/src/org/graalvm/compiler/truffle/compiler/benchmark/pelang/PELangExpressionNode.java]: Yeah you are right, I will definitively do this.

**Comment by boris-spas** [comment]
Is there still interest in merging this PR?
If the conflicts are resolved I'd try to shepherd it.

**Comment by flortsch** [comment]
I think this PR can be closed for now. We did some bigger changes in the language itself and the benchmarking classes. We are also working on some partial evaluation improvements in general. So I think it's better when we come up with a new PR in the future.

**Pull request closed without merging**